kythe/java/com/google/devtools/kythe/extractors/java/bazel/aspect.bzl

```diff
@@ -0,0 +1,136 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("//tools/build_rules/verifier_test:verifier_test.bzl", "extract")
+
+_mnemonic = "Javac"
+
+def _extract_java(target, ctx):
+    if JavaInfo not in target or not hasattr(ctx.rule.attr, "srcs"):
+        return None
+
+    javac_action = None
+    for a in target.actions:
+        if a.mnemonic == _mnemonic:
+            javac_action = a
+            break
+
+    if not javac_action:
+        return None
+
+    kzip = ctx.actions.declare_file(ctx.label.name + ".xa.java.kzip")
+
+    info = target[JavaInfo]
+    compilation = info.compilation_info
+    annotations = info.annotation_processing
+
+    source_files = []
+    for src in ctx.rule.files.srcs:
+        source_files += [src.path]
+
+    classpath = [j.path for j in compilation.compilation_classpath.to_list()]
+    bootclasspath = [j.path for j in compilation.boot_classpath]
+
+    processorpath = []
+    processors = []
+    if annotations and annotations.enabled:
+        processorpath += [j.path for j in annotations.processor_classpath.to_list()]
+        processors = annotations.processor_classnames
+
+    output_jar = []
+    for jar in info.outputs.jars:
+        output_jar += [jar.class_jar.path]
+
+    # TODO(schroederc): sourcegendir
+    xa = struct(**{
+        "owner": str(target.label),
+        "mnemonic": _mnemonic,
+        "[blaze.JavaCompileInfo.java_compile_info]": struct(
+            outputjar = output_jar,
+            classpath = classpath,
+            source_file = source_files,
+            javac_opt = compilation.javac_options,
+            processor = processors,
+            processorpath = processorpath,
+            bootclasspath = bootclasspath,
+        ),
+    })
+    textXA = ctx.actions.declare_file(ctx.label.name + ".xa.textproto")
+    ctx.actions.write(
+        output = textXA,
+        content = xa.to_proto(),
+    )
+
+    xa = ctx.actions.declare_file(ctx.label.name + ".xa")
+    xaArgs = ctx.actions.args()
+    xaArgs.add_all([textXA, xa])
+    ctx.actions.run(
+        outputs = [xa],
+        inputs = [textXA],
+        arguments = [xaArgs],
+        executable = ctx.executable._write_extra_action,
+    )
+
+    extractArgs = ctx.actions.args()
+    extractArgs.add_all([xa, kzip, ctx.file._java_aspect_vnames_config])
+    deps = [javac_action.inputs]
+    ctx.actions.run(
+        outputs = [kzip],
+        inputs = depset([xa, ctx.file._java_aspect_vnames_config], transitive = deps),
+        executable = ctx.executable._java_bazel_extractor,
+        arguments = [extractArgs],
+    )
+
+    return kzip
+
+def _extract_java_aspect(target, ctx):
+    kzip = _extract_java(target, ctx)
+    if not kzip:
+        return struct()
+    return [OutputGroupInfo(kzip = [kzip])]
+
+# Aspect to run the Bazel Javac extractor on all specified Java targets.
+extract_java_aspect = aspect(
+    _extract_java_aspect,
+    attr_aspects = ["srcs"],
```

Comment:
This will cause the aspect to propagate recursively down all of the rules specified in the targets `srcs`.  I'm not sure that's what you want, though.